File: Searching and Sorting Algorithms/18CountingSort.py
def counting_sort(arr):
    if not arr:
        return
    max_val=max(arr)
    min_val=min(arr)
    width=max_val-min_val+1
    count=[0]*(width)

    for el in arr:
        index=el-min_val
        count[index]+=1
    
    for i in range(1,width):
        count[i]+=count[i-1]
    print(count)
    n=len(arr)
    res=[0]*(n)
    
    for i in range(n-1,-1,-1):
        num=arr[i]
        index=num-min_val
        count[index]-=1
        res[count[index]]=num
    return res
    
arr=[4,2,2,8,3,3,1]
print(counting_sort(arr))

File: Searching and Sorting Algorithms/codigo_cs/18CountingSort.cs
using System;

public class Test18
{
    public static void CountingSort(int[] arr)
    {
        int minVal = arr.Min();
        int maxVal = arr.Max();
        int width = maxVal - minVal + 1;
        int[] count = new int[width];
        foreach (int el in arr)
        {
            int index = el - minVal;
            count[index] += 1;
        }
        for (int i = 1; i < width; i++)
        {
            count[i] += count[i - 1];
        }
        int n = arr.Length;
        int[] res = new int[n];
        for (int i = n - 1; i >= 0; i--)
        {
            int num = arr[i];
            int index = num - minVal;
            count[index] -= 1;
            res[count[index]] = num;
        }

        // for (int i = 0; i < n; i++)
        // {
        //     arr[i] = res[i];
        // }
        Console.WriteLine(string.Join(" ", res));





    }
}